src/rust/engine/grpc_util/src/retry.rs

```diff
@@ -0,0 +1,63 @@
+// Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+use std::time::Duration;
+
+use futures::Future;
+use rand::{thread_rng, Rng};
+use tonic::{Code, Response, Status};
+
+fn is_retryable(status: &Status) -> bool {
+  matches!(
+    status.code(),
+    Code::Aborted
+      | Code::Cancelled
+      | Code::Internal
+      | Code::ResourceExhausted
+      | Code::Unavailable
+      | Code::Unknown
+  )
+}
+
+/// Retry a gRPC client operation using exponential back-off to delay between attempts.
+#[inline]
+pub async fn retry_call<T, C, F, Fut>(client: C, f: F) -> Result<Response<T>, Status>
+where
+  C: Clone,
+  F: Fn(C) -> Fut,
+  Fut: Future<Output = Result<Response<T>, Status>>,
+{
+  const INTERVAL_DURATION: Duration = Duration::from_millis(10);
+  const MAX_RETRIES: u32 = 3;
+  const MAX_BACKOFF_DURATION: Duration = Duration::from_secs(5);
```

Comment:
Should these be configurable?

Reply:
I don't think so, at least to start. Imo there's a risk of options fatigue. For example, how is an end user supposed to determine what values make sense? These are finicky to play with.

Another important detail here: remote cache failures should not break Pants. If these numbers aren't perfectally optimal, that's not a big deal.

Reply:
I could buy `MAX_RETRIES` being configurable (people may want different trade-offs there), but I can't really justify the others.

Reply:
I'll keep them non-configurable for now.